src/foam/nanos/ruler/RuleDebugger.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.ruler',
+  name: 'RuleDebugger',
+  extends: 'foam.nanos.ruler.Rule',
+  documentation: `Rule to debug and print out all other rules.
+  INSTRUCTIONS FOR USE:
+  DAOKey: the DAO this rule runs on and subsequently tests.
+  After: set to true to only test after rules. default is only test before rules
+  Operation: to filter debugging rules to only certain operation. default is 3 (Create and Update)
+  Rule Group: Do not forget to set one, and ensure it is on for debugging.
+
+  Rule will run and log relevant predicate and permission evaluation based on the provided obj, on the operation and dao that is specified.
+  `,
+
+  implements: [
+    'foam.nanos.ruler.RuleAction'
+  ],
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.X',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.nanos.logger.Logger',
+    'foam.nanos.auth.AuthService',
+    'foam.nanos.ruler.Rule',
+    'foam.nanos.ruler.RuleGroup',
+    'foam.nanos.auth.User',
+    'static foam.mlang.MLang.EQ',
+    'static foam.mlang.MLang.NEQ',
+    'static foam.mlang.MLang.AND',
+    'java.util.List'
+  ],
+
+  properties: [
+    {
+      name: 'action',
+      transient: true,
+      visibility: 'HIDDEN',
+      javaGetter: 'return this;'
+    },
+    {
+      name: 'predicate',
+      transient: true,
+      visibility: 'HIDDEN',
+      javaGetter: 'return foam.mlang.MLang.TRUE;'
+    },
+    {
+      class: 'Int',
+      name: 'priority',
+      value: 2147483647,
+      documentation: 'Ensure this rule always runs first'
+    },
+    {
+      class: 'Enum',
+      of: 'foam.nanos.dao.Operation',
+      name: 'operation',
+      value: 'CREATE_OR_UPDATE'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        x = getX();
```

Comment:
the first arg of applyAction is x already, doesn't it work for you?

Reply:
I need system X to find all rules